keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,372 @@
+ - Audit logs can't be associated to logs related to user operation (kubectl operation), because auditID is different for each http request.
+
+#### Story 2
+
+Failed to attach PV to pod
+Prerequisite: It has been confirmed that the PV has been created successfully.
+In this case, the volume generation on the storage side is OK, and there is a possibility that the mount process to the container in the pod is NG.
+In order to identify the cause, it is necessary to look for the problem area while checking the component (kubelet) log as well as the system side syslog and mount related settings.
+
+This log tracking feature is useful to identify the logs related to specific user operation  and cluster processing, and can reduce investigation cost in such cases.
+### Summary of Cases
+
+ - Given a component log(such as error log), find the API request that caused this (error) log.
+ - Given an API Request(such as suspicious API request), find the resulting component logs.
+
+### Logging metadata
+
+We use three logging meta-data, and propagate them each K8s component by using OpenTelemetry.
+OpenTelemetry has SpanContext and [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) which is used for propagation of K8s component.
+
+| meta-data name | feature |
+| ------ | ------ |
+| traceid | We use SpanContext.TraceID as traceid<br>traceid spans an user request.<br>traceid is unique for user's request |
+| spanid | We use SpanContext.SpanID as spanid<br>spanid spans a controller action.<br>spanid is unique for controller action |
+| initialtraceid | We implement new id(InitialTraceID) to SpanContext<br>We use [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) to propagate initialtraceid<br/>We use UID of root object as initialtraceid<br>initialtraceid spans the entire object lifecycle. <br>initialtraceid is unique for related objects |
+
+All of three id's inception is from object creation and it dies with object deletion
+
+## Design Details
+
+### Prerequisite
+
+We need to consider three cases:
+- Case1: Requests from kubectl that creating an object
+- Case2: Requests from kubectl other than creating (e.g. updating, deleting) an object
+- Case3: Requests from controllers
+
+The design below is based on the above three cases
+
+The following picture show our design![design](./overview.png)
+
+
+
+we don't have any modifications in kubectl in this design.
+
+### API Server Tracing
+
+**1. refer to [KEP647](https://github.com/kubernetes/enhancements/issues/647)**
+
+KEP647 will help to
+
+```
+Goals
+The API Server generates and exports spans for incoming and outgoing requests.
+The API Server propagates context from incoming requests to outgoing requests.
+```
+
+**2. [Make up a new outgoing request](#design-of-id-progagation-in-client-go)**
+
+### Design of ID propagation (controller)
```

Comment:
I'd like this split into a few sections:

1. New context propagation library functions: 
  a. `func WithObject(ctx context.Context, meta metav1.ObjectMeta) context.Context`
2. Example controller usage, including:
  a. How to log with the metadata from a context.Context.
  b. How to use the context.Context in outgoing calls, and what impact this would have.

Reply:
thanks for your comments, i'm not sure if if have got all your point. i update the controller section like below. feel free to let me know if i got something wrong.

> New context propagation library functions:
> a. func WithObject(ctx context.Context, meta metav1.ObjectMeta) context.Context

refer to **1. New context propagation library function**

> Example controller usage, including:
>a. How to log with the metadata from a context.Context.

refer to **2. How to log the metadata from a context.Context**

> b. How to use the context.Context in outgoing calls, and what impact this would have.

refer to **3. Propagate Span and Baggage with context.Context across object**


**1. New context propagation library function**

```
func SpanContextWithObject(ctx context.Context, meta metav1.Object) context.Context
```
SpanContextWithObject returns a context including  Span and Baggage according to the passed object, it do something like below pseudo code:
```
func SpanContextWithObject(ctx context.Context, meta metav1.Object) context.Context{
	// 1. extract baggage from annotation and chain it to ctx
	// 2. decode annotation to SpanContext
	// 3. generate Span from SpanContext
	// 4. chain Span to ctx which is required by propagators.Inject
}
```
**2. How to log the metadata from a context.Context**

Below is an example to log the metadata

```
spanctx := apitrace.SpanContextFromContext(ctx)
initTraceID := otel.BaggageValue(ctx, "Initial-Trace-Id").AsString()
klog.V(2).Infof("InitialTraceID: %s, TraceID: %s, SpanID: %s\n", initTraceID, spanctx.TraceID, spactx.SpanID)
```
Please note that changing the log is not the goal of this KEP.

**3. Propagate Span and Baggage with context.Context across object**

When controllers create/update/delete an object A based on another B, we propagate context from B to A. Below is an example to show how deployment propagate Span and Baggage to replicaset.

```diff
 func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, oldRSs []*apps.ReplicaSet, createIfNotExisted bool) (*apps.ReplicaSet, error) {
+       ctx := httptrace.SpanContextWithObject(context.Background(), d)
        existingNewRS := deploymentutil.FindNewReplicaSet(d, rsList)
@@ -220,7 +227,8 @@ func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, old
        // hash collisions. If there is any other error, we need to report it in the status of
        // the Deployment.
        alreadyExists := false
-       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(context.TODO(), &newRS, metav1.CreateOptions{})
+       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(ctx, &newRS, metav1.CreateOptions{})
```
In order to propagate context across all objects we concerned , we also need to change some APIs by adding `ctx context.Context` to parameter list whose parameters doesn't contain context.Context yet. Below APIs will be impacted so far.

| APIs                          | file name                                                    |
| ----------------------------- | ------------------------------------------------------------ |
| createPods()                  | pkg/controller/controller_utils.go                           |
| CreatePodsWithControllerRef() | pkg/controller/controller_utils.go<br />pkg/controller/replication/conversion.go<br />pkg/controller/daemon/daemon_controller.go<br />pkg/controller/replication/conversion.go |

Reply:
> New context propagation library functions:
> a. func WithObject(ctx context.Context, meta metav1.ObjectMeta) context.Context

Would attaching a logr instance which then contains that object as an additional key/value item also work? It's a more flexible.

The downside is that code then has to be converted to structured logging (but that's the goal anyway, right?) and that the object cannot be retrieved anymore separately (which may or may not be relevant).

Reply:
Thanks for your comment. I want to know more details of your suggestion. Maybe I could not understand it...
Do  you mean that we should add logr instance to the library function?(Like, `func WithObject(ctx context.Context, meta metav1.ObjectMeta, l logr.Logger) context.Context`.)

The final goal is output the tracking information using structured logging format. However, I assume that it is out of scope in this KEP. Of course, this KEP and implementation should be with structured logging in mind.
I'm sorry if this question is off the mark.

Reply:
The goal is to to have something like https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/log.go in klog (perhaps as klog/kcontext/kcontext.go?), except that the `FromContext` function in klog falls back to the global klog instance wrapped by klogr, so a caller can always be sure that it gets a logger and thus doesn't need to check for nil.

Then gradually packages can be converted from logging via klog to logging via a logr instance that they get with that `FromContext` function -> structured logging, with additional information attached to each log message by the caller. As an added bonus, the caller can decide to raise the log level for the call.

> The final goal is output the tracking information using structured logging format. However, I assume that it is out of scope in this KEP.

That depends. If you are proposing that each log site should do something like "retrieve objects, log with klog.InfoS", then I think we should consider whether that is the best solution for the problem. We don't want to propose one solution now (this KEP) and then another solution later (a KEP for passing a logger via the context).

As @dashpole said, it will be easier to discuss when you show how the proposed enhancement will be used by code in Kubernetes.

Reply:
According to your comments by far, I feel that you mainly concern that how we modify the logs with the feature proposed in this KEP right?

I just want to clarify that In our initial vision "how we change logs" is out of this KEP's scope. With David's earlier suggestion we decided to split it down, to just focus on "bring k8s a new feature which could generate and propagate unique IDs across different components and objects" first, and then propose to change logs after that, because it requires more consideration and discussion with sigs.

But I realized that what you commented is worth considering, we will show some examples to answer your points, thanks.

Reply:
> Some thoughts:
> 

thanks for you inputs.

> * As far as I can see, this KEP currently has no details about how that annotation would look.

Our annotation there looks like:
```
Annotations:    
   trace.kubernetes.io.initial: 9ee7aeb7-46c6-4e67-9b7b-9fb10d7354ac (an unique ID, write once, read only later)
   trace.kubernetes.io.span.context: KA5YXmPdf69mdiTADM9jYutbzITJmz9MAA== (it's encoded, 00-280e585e63dd7faf667624c00ccf6362-c9f8cb4ab85acd63-00 after decoding)
```